fuse_core/src/callback_wrapper.cpp

```diff
@@ -0,0 +1,148 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2022, Brett Downing
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <fuse_core/callback_wrapper.h>
+
+namespace fuse_core
+{
+
+
+  CallbackAdapter::CallbackAdapter(std::shared_ptr<rclcpp::Context> context_ptr){
+
+    rcl_guard_condition_options_t guard_condition_options =
+        rcl_guard_condition_get_default_options();
+
+    // Guard condition is used by the wait set to handle execute-or-not logic
+    gc_ = rcl_get_zero_initialized_guard_condition();
+    if (RCL_RET_OK != rcl_guard_condition_init(
+      &gc_, context_ptr->get_rcl_context().get(), guard_condition_options)
+    ) {
+      throw std::runtime_error("Could not init guard condition for callback waitable.");
+    }
+  }
+
+  /**
+   * @brief tell the CallbackGroup how many guard conditions are ready in this waitable
+   */
+  size_t CallbackAdapter::get_number_of_ready_guard_conditions() { return 1;}
+
+
+  /**
+   * @brief tell the CallbackGroup that this waitable is ready to execute anything
+   */
+  bool CallbackAdapter::is_ready(rcl_wait_set_t * wait_set) {
+    (void) wait_set;
+
+    // Prevent multiple simultaneous calls of is_ready (causing the executor to wake multiple times)
+    std::lock_guard<std::mutex> lock(ready_mutex_);
```

Comment:
My earlier comment about the executor waking multiple times in `is_ready()` doesn't make sense to me today :no_mouth: 

This should be `queue_mutex_`, instead of `ready_mutex_`, but maybe no-mutex is OK after all, assuming the executor won't call `is_ready()` again until after it calls `take_data()`.

Reply:
habitually locking data structures (queue) that might be modified by another thread is good practices,
but in this case it can still report work is ready to be done when another thread has previously checked and is blocked waiting to pop the last job; so it didn't seem worth locking.

Reply:
https://github.com/locusrobotics/fuse/pull/284/commits/cdeca18dee79f9a50a3788c844308a84f192b717